src/utils/series.ts

```diff
@@ -0,0 +1,105 @@
+import { addQueryParams, slugify } from './formatting';
+
+import type { PlaylistItem } from '#types/playlist';
+import type { EpisodeWithSeason, Season, Series } from '#types/series';
+
+export const episodeURL = (series: Series, episodeId?: string, play: boolean = false, playlistId?: string | null) =>
+  addQueryParams(`/s/${series.series_id}/${slugify(series.title)}`, {
+    e: episodeId,
+    r: playlistId,
+    play: play ? '1' : null,
+  });
+
+export const generateSeriesMetadata = (seriesPlaylist: Series) => {
+  const seriesCanonical = `${window.location.origin}${episodeURL(seriesPlaylist)}`;
+
+  return {
+    '@type': 'TVSeries',
+    '@id': seriesCanonical,
+    name: seriesPlaylist.title,
+    numberOfEpisodes: seriesPlaylist.episode_count,
+    numberOfSeasons: seriesPlaylist.seasons.length,
+  };
+};
+
+export const generateEpisodeJSONLD = (seriesPlaylist: Series, episode: PlaylistItem) => {
+  const episodeCanonical = `${window.location.origin}${episodeURL(seriesPlaylist, episode.mediaid)}`;
+  const seriesMetadata = generateSeriesMetadata(seriesPlaylist);
+
+  return JSON.stringify({
+    '@context': 'http://schema.org/',
+    '@type': 'TVEpisode',
+    '@id': episodeCanonical,
+    episodeNumber: episode.episodeNumber,
+    seasonNumber: episode.seasonNumber,
+    name: episode.title,
+    uploadDate: secondsToISO8601(episode.pubdate),
+    partOfSeries: seriesMetadata,
+  });
+};
+
+/** We need to add episodeNumber and seasonNumber to each media item we got via watchlists
+ *  That will help with further data retrieval
+ */
+export const enrichMediaItems = (series: Series | undefined, mediaItems: PlaylistItem[] | undefined): PlaylistItem[] => {
+  const episodes = (series?.seasons || []).reduce((acc: EpisodeWithSeason[], season: Season) => {
+    const episodes = season.episodes.map((episode) => ({ ...episode, season_number: season.season_number }));
+
+    return [...acc, ...episodes];
+  }, []);
+
+  const itemsWithEpisodes = (mediaItems || []).map((item) => {
+    const episode = episodes.find((episode) => episode.media_id === item.mediaid) as EpisodeWithSeason;
+
+    return { ...item, seasonNumber: String(episode.season_number), episodeNumber: String(episode.episode_number) };
+  });
+
+  return itemsWithEpisodes;
+};
+
+/** Next episode to show after video is complete */
+export const getNextEpisode = (series: Series, seasonNumber: number, media: PlaylistItem) => {
+  const episodes = series.seasons[seasonNumber - 1].episodes;
+
+  const episodeIndex = episodes.findIndex((el) => el.media_id === media.mediaid);
+
+  // If we have the last episode in the last season => we leave the same video
+  if (series.seasons.length === seasonNumber && episodeIndex === episodes.length - 1) {
+    return media.mediaid;
+  }
+
+  // If we have the last episode in the current season => we change the season
+  if (episodes.length - 1 === episodeIndex) {
+    return series.seasons[seasonNumber + 1].episodes[0].media_id;
```

Comment:
Potentially no episodes in a season?

```suggestion
    return series.seasons[seasonNumber + 1].episodes[0]?.media_id;
```